app/ui/src/Main.elm

```diff
@@ -1,64 +1,53 @@
-import Html exposing (div, button, text, input)
+module Main exposing (..)
+
+import Html
 import Html.App as App
-import Html.Events exposing (onClick, onInput)
-import Task
-import Http
-import Json.Decode as Json
+import Server as API
+import Model exposing (Model, Output(..))
+import Message exposing (Message(..))
+import Components as UI
 
 
 main =
-  App.program { init = model, view = view, update = update, subscriptions = subscriptions }
-
-type alias Model =
-  { text : String
-  , finalText : String
-  }
-
-model : (Model, Cmd a)
-model =
-  (Model "" "", Cmd.none)
-
-
-view model =
-  div []
-    [ input [ onInput UpdateText ] []
-    , button [ onClick GetMoreData ] [ text "Update" ]
-    , div [] [ text (toString model.finalText) ]
-    ]
+    App.program
+        { init = init
+        , update = update
+        , subscriptions = subscriptions
+        , view = view
+        }
 
 
-type Msg = UpdateText String | UpdateFinalText String | GetMoreData | GetDataFail Http.Error
+init : ( Model, Cmd a )
+init =
+    ( Model "" (Single "Ready."), Cmd.none )
 
-update : Msg -> Model -> (Model, Cmd Msg)
-update msg model =
-  case msg of
-    UpdateText text ->
-      ({ model | text = text }, Cmd.none)
 
-    UpdateFinalText finalText ->
-      ({ model | finalText = finalText }, Cmd.none)
+update : Message -> Model -> ( Model, Cmd Message )
+update msg { query, output } =
+    case msg of
+        ChangeQuery text ->
+            ( Model text output, Cmd.none )
 
-    GetMoreData ->
-      (model, getData model.text)
+        Fetch ->
+            ( Model query (Model.Single "Awaiting response..."), API.send query )
 
-    GetDataFail error ->
-      ({ model | finalText = (toString error) }, Cmd.none)
+        Receive json ->
+            case Model.fromJson json of
+                Err err ->
+                    ( Model query (Model.Single err), Cmd.none )
```

Comment:
I find it easier to read using the literal syntax than the Model constructor function as the order of the arguments no longer matter.